### DISCUSSION TITLE
Sending a file - POST /translateSubtitle HTTP/1.1 -  422 Unprocessable Entity

### DISCUSSION THREAD

**Discussion by chen-rn** [fastapi/fastapi#9731] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
//Here is my main.py
from helpers.translate_srt import translate_srt

from fastapi import FastAPI, File, UploadFile, HTTPException
import traceback

app = FastAPI()
import logging


@app.post("/translateSubtitle")
async def translate_subtitle(target_language: str, file: UploadFile = File(...)):
    try:
        # Log the received request data
        logging.info(
            f"Received request to translate subtitle. Target language: {target_language}"
        )

        # Read the content of the file
        file_content = await file.read()

        # Log the content of the file
        logging.info(
            f"Received file with content: {file_content[:100]}..."
        )  # Log only the first 100 characters for brevity

        # Placeholder translation function
        translated_content = translate_srt(file_content, target_language)

        # Log the success of the translation process
        logging.info(f"Successfully translated subtitle to {target_language}")

        # Return the translated content in the response
        return {"file_content": translated_content}
    except Exception as e:
        # Log the exception
        logging.error(
            f"An error occurred while translating subtitle to {target_language}: {e}"
        )

        traceback_str = "".join(
            traceback.format_exception(etype=type(e), value=e, tb=e.__traceback__)
        )
        raise HTTPException(
            status_code=400,
            detail={"error": "An error occurred.", "traceback": traceback_str},
        )



//I'm calling it from a node server, and here is my typescript code
export async function translateSubtitle(
	filePath,
	targetLanguage,
	outputFilePath,
) {
	// Create a new FormData instance
	const formData = new FormData()

	// Append the file and target language to the form
	formData.append("target_language", targetLanguage)
	formData.append("file", fs.createReadStream(filePath))

	try {
		// Make the POST request to the FastAPI server
		const response = await axios.post(
			process.env.TRANSLATION_ENDPOINT,
			formData,
			{
				headers: formData.getHeaders(),
			},
		)

		// Check if the response data is a string before writing to file
		if (typeof response.data === "string" || response.data instanceof String) {
			// Convert response.data to a string primitive if it's a String object
			const data = response.data.toString()

			// Write the translated subtitles to the output file
			fs.writeFile(outputFilePath, data, (err) => {
				if (err) {
					console.error(`Error writing to file: ${err}`)
				} else {
					console.log(
						`Successfully wrote translated subtitles to ${outputFilePath}`,
					)
				}
			})
		} else {
			console.error("Received data is not a string.")
		}
	} catch (error) {
		console.error(`Error: ${error}`)
		console.error(`Stack: ${error.stack}`)
	}
}
```

### Description

I am trying to send a SRT file along with the POST request through a FormData.

However,  I am getting an error that states: "POST /translateSubtitle HTTP/1.1" 422 Unprocessable Entity

This is my first time setting up a FastAPI server so I'm not too sure how to proceed from here.
It seems that FormData is the default way of sending files over HTTP?

Any guidance as to how I can approach debugging this is appreciated!

Note: this is running in virtual python environment, idk if that matters!?

### Operating System

macOS

### FastAPI Version

0.65.2

### Python Version

Python 3.11.4

**Comment by arjwilliams** [comment, marked as answer]
Hi @chen-rn,

You need to change your `target_language` parameter to be recognised as part of the form data:

``` diff
- from fastapi import FastAPI, File, UploadFile, HTTPException
+ from fastapi import FastAPI, File, UploadFile, HTTPException, Form

- async def translate_subtitle(target_language: str, file: UploadFile = File(...)):
+ async def translate_subtitle(target_language: str = Form(...), file: UploadFile = File(...)):
```

Having it as `target_language: str` means FastAPI expects it to be a query parameter, rather than part of the body.

See https://fastapi.tiangolo.com/tutorial/request-forms/#__tabbed_1_3 for more info.

**Reply by chen-rn** [reply]
Side note, I'm not too familiar with the "convention" of these things.

Should I keep it in the query params and just change my axios request?
Or is it better to put it both in body?

**Reply by arjwilliams** [reply]
If you have a GET route on the FastAPI side where you return the untranslated SRT files, then I would add the translation functionality in there, and have the target_language as a query parameter.

But if that's not possible (ie. if the SRT files live on the node side), then passing the language in the body of the POST request is fine.

**Comment by chen-rn** [comment]
Ohhhh! Gotcha, thank you so much!
